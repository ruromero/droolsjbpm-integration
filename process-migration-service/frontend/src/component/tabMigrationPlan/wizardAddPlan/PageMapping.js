import React, { Component } from "react";
import PropTypes from "prop-types";
import { Button } from "patternfly-react";

import PageMappingDiagrams from "./PageMappingDiagrams";
import PageMappingDropdownNode from "./PageMappingDropdownNode";
import Notification from "../../Notification";
import { ALERT_TYPE_ERROR } from "patternfly-react/dist/js/components/Alert/AlertConstants";
import { Form } from "patternfly-react/dist/js/components/Form";
import Col from "patternfly-react/dist/js/components/Grid/Col";
import FormGroup from "patternfly-react/dist/js/components/Form/FormGroup";
import ControlLabel from "patternfly-react/dist/js/components/Form/ControlLabel";
import Label from "patternfly-react/dist/js/components/Label/Label";

export default class PageMapping extends Component {
  constructor(props) {
    super(props);
    this.state = {
      errorMsg: ""
    };
    // The mappings page doesn't require any validation
    this.props.onIsValid(true);
  }

  handleSourceDropdownChange = option => {
    this.setState({
      sourceNodeTitle: `${option.name} (${option.type})`,
      sourceNode: option,
      sourcePreviousNode: this.state.sourceCurrentNode,
      sourceCurrentNode: option.id
    });
  };

  handleTargetDropdownChange = option => {
    this.setState({
      targetNodeTitle: `${option.name} (${option.type})`,
      targetNode: option,
      targetPreviousNode: this.state.targetCurrentNode,
      targetCurrentNode: option.id
    });
  };

  handleMapButtonClick = () => {
    if (
      this.state.sourceNode !== undefined &&
      this.state.targetNode !== undefined
    ) {
      if (this.state.sourceNode.type !== this.state.targetNode.type) {
        this.setState({
          errorMsg: "Source and Target nodes must be of the same type"
        });
        return;
      }
      let mappings = this.props.plan.mappings;
      if (mappings === undefined) {
        mappings = {};
      }
      if (mappings[this.state.sourceNode.id]) {
        this.setState({
          errorMsg: "You cannot map the same source node multiple times"
        });
        return;
      }
      mappings[this.state.sourceNode.id] = this.state.targetNode.id;
      this.props.onMappingsChange(mappings);
    }
  };

  deleteMapping = id => {
    const mappings = this.props.plan.mappings;
    delete mappings[id];
    this.props.onMappingsChange(mappings);
  };

  buildMappings = () => {
    const mappingValues = [];
    const mappings = this.props.plan.mappings;
    if (mappings === undefined || mappings === {}) {
      return mappingValues;
    }
    Object.keys(mappings).forEach(key => {
      const filteredSource = this.props.sourceProcess.nodes.filter(
        node => node.id === key
      );
      let sourceNode;
      if (filteredSource.length === 1) {
        sourceNode = filteredSource[0];
      }
      const filteredTarget = this.props.targetProcess.nodes.filter(
        node => node.id === mappings[key]
      );
      let targetNode;
      if (filteredTarget.length === 1) {
        targetNode = filteredTarget[0];
      }
      if (sourceNode && targetNode) {
        mappingValues.push({
          id: sourceNode.id,
          label: sourceNode.name + ">" + targetNode.name
        });
      }
    });
    return mappingValues;
  };

  buildMappingLabels = () =>
    this.buildMappings().map(mapping => (
      <Label
        key={mapping.id}
        style={{ margin: "5px 5px 0 0", display: "inline-block" }}
        onRemoveClick={() => this.deleteMapping(mapping.id)}
      >
        {mapping.label}
      </Label>
    ));

  render() {
    const notification = (
      <Notification
        type={ALERT_TYPE_ERROR}
        message={this.state.errorMsg}
        onDismiss={() => this.setState({ errorMsg: "" })}
      />
    );
    if (
      this.props.sourceProcess.processId !== undefined &&
      this.props.targetProcess.processId !== undefined
    ) {
      return (
        <Form horizontal>
          {this.state.errorMsg && notification}
          <FormGroup controlId="mappingHeader">
            <Col sm={4} componentClass={ControlLabel}>
              Source: {this.props.sourceProcess.containerId} {" / "}
              {this.props.sourceProcess.processId}
            </Col>
            <Col sm={4} componentClass={ControlLabel}>
              Target: {this.props.targetProcess.containerId} {" / "}
              {this.props.targetProcess.processId}
            </Col>
            <Col sm={4} />
          </FormGroup>
          <FormGroup controlId="mappingDropdowns">
            <Col sm={4} className="text-right">
              <PageMappingDropdownNode
                options={this.props.sourceProcess.nodes}
                title={
                  this.state.sourceNodeTitle
                    ? this.state.sourceNodeTitle
                    : "Source Nodes"
                }
                onDropdownChange={this.handleSourceDropdownChange}
              />
            </Col>
            <Col sm={4} className="text-right">
              <PageMappingDropdownNode
                options={this.props.targetProcess.nodes}
                title={
                  this.state.targetNodeTitle
                    ? this.state.targetNodeTitle
                    : "Target Nodes"
                }
                onDropdownChange={this.handleTargetDropdownChange}
              />
            </Col>
            <Col sm={4}>
              <Button bsStyle="primary" onClick={this.handleMapButtonClick}>
                Map these two nodes
              </Button>
            </Col>
          </FormGroup>
          <FormGroup controlId="mappingLabels">
            <Col sm={4} componentClass={ControlLabel}>
              Mappings:
            </Col>
            <Col sm={8}>{this.buildMappingLabels()}</Col>
          </FormGroup>
          <FormGroup controlId="mappingDiagrams">
            <PageMappingDiagrams
              sourceCurrentNode={this.state.sourceCurrentNode}
              sourcePreviousNode={this.state.sourcePreviousNode}
              targetCurrentNode={this.state.targetCurrentNode}
              targetPreviousNode={this.state.targetPreviousNode}
              sourceProcess={this.props.sourceProcess}
              targetProcess={this.props.targetProcess}
            />
          </FormGroup>
        </Form>
      );
    } else {
      //no process info retrieved from backend yet, just display an empty tag to avoid error.
      return <div />;
    }
  }
}

PageMapping.propTypes = {
  plan: PropTypes.object.isRequired,
  sourceProcess: PropTypes.object.isRequired,
  targetProcess: PropTypes.object.isRequired,
  onMappingsChange: PropTypes.func.isRequired,
  onIsValid: PropTypes.func.isRequired
};
