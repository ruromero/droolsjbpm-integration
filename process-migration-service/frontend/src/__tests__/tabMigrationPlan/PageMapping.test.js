import renderer from "react-test-renderer";
import React from "react";
import PageMapping from "../../component/tabMigrationPlan/wizardAddPlan/PageMapping";
import evaluationDef from "../../../mock_data/process_definition_evaluation.json";
import mortgageDef from "../../../mock_data/process_definition_mortgage.json";

test("PageMapping renders correctly using snapshot", () => {
  const plan = {};
  const mockFn = jest.fn();
  const tree = renderer
    .create(
      <PageMapping
        plan={plan}
        sourceProcess={evaluationDef}
        targetProcess={mortgageDef}
        onMappingsChange={mockFn}
        onIsValid={mockFn}
      />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
