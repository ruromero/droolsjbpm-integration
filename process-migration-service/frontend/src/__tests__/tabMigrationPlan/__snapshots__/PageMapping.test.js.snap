// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PageMapping renders correctly using snapshot 1`] = `
<form
  className="form-horizontal"
>
  
  <div
    className="form-group"
  >
    <label
      className="col-sm-4 control-label"
      htmlFor="mappingHeader"
    >
      Source: 
      evaluation_2.0.0-SNAPSHOT
       
       / 
      evaluation
    </label>
    <label
      className="col-sm-4 control-label"
      htmlFor="mappingHeader"
    >
      Target: 
      mortgage-process_1.0.0-SNAPSHOT
       
       / 
      Mortgage_Process.MortgageApprovalProcess
    </label>
    <div
      className="col-sm-4"
    />
  </div>
  <div
    className="form-group"
  >
    <div
      className="text-right col-sm-4"
    >
      <div
        className="dropdown btn-group"
      >
        <button
          aria-expanded={false}
          aria-haspopup={true}
          className="dropdown-toggle btn btn-default"
          disabled={false}
          id="nodeMappingDropdown"
          onClick={[Function]}
          onKeyDown={[Function]}
          role="button"
          type="button"
        >
          Source Nodes
           
          <span
            className="caret"
          />
        </button>
        <ul
          aria-labelledby="nodeMappingDropdown"
          className="dropdown-menu"
          role="menu"
        >
          <li
            className=""
            role="presentation"
          >
            <a
              href="#"
              onClick={[Function]}
              onKeyDown={[Function]}
              role="menuitem"
              tabIndex="-1"
            >
              HR Evaluation
               (
              userTask
              )
            </a>
          </li>
          <li
            className=""
            role="presentation"
          >
            <a
              href="#"
              onClick={[Function]}
              onKeyDown={[Function]}
              role="menuitem"
              tabIndex="-1"
            >
              Input
               (
              userTask
              )
            </a>
          </li>
          <li
            className=""
            role="presentation"
          >
            <a
              href="#"
              onClick={[Function]}
              onKeyDown={[Function]}
              role="menuitem"
              tabIndex="-1"
            >
              PM Evaluation
               (
              userTask
              )
            </a>
          </li>
          <li
            className=""
            role="presentation"
          >
            <a
              href="#"
              onClick={[Function]}
              onKeyDown={[Function]}
              role="menuitem"
              tabIndex="-1"
            >
              Self Evaluation
               (
              userTask
              )
            </a>
          </li>
        </ul>
      </div>
    </div>
    <div
      className="text-right col-sm-4"
    >
      <div
        className="dropdown btn-group"
      >
        <button
          aria-expanded={false}
          aria-haspopup={true}
          className="dropdown-toggle btn btn-default"
          disabled={false}
          id="nodeMappingDropdown"
          onClick={[Function]}
          onKeyDown={[Function]}
          role="button"
          type="button"
        >
          Target Nodes
           
          <span
            className="caret"
          />
        </button>
        <ul
          aria-labelledby="nodeMappingDropdown"
          className="dropdown-menu"
          role="menu"
        >
          <li
            className=""
            role="presentation"
          >
            <a
              href="#"
              onClick={[Function]}
              onKeyDown={[Function]}
              role="menuitem"
              tabIndex="-1"
            >
              Correct Data
               (
              userTask
              )
            </a>
          </li>
          <li
            className=""
            role="presentation"
          >
            <a
              href="#"
              onClick={[Function]}
              onKeyDown={[Function]}
              role="menuitem"
              tabIndex="-1"
            >
              Final Approval
               (
              userTask
              )
            </a>
          </li>
          <li
            className=""
            role="presentation"
          >
            <a
              href="#"
              onClick={[Function]}
              onKeyDown={[Function]}
              role="menuitem"
              tabIndex="-1"
            >
              Increase Down Payment
               (
              userTask
              )
            </a>
          </li>
          <li
            className=""
            role="presentation"
          >
            <a
              href="#"
              onClick={[Function]}
              onKeyDown={[Function]}
              role="menuitem"
              tabIndex="-1"
            >
              Mortgage Calculation
               (
              businessRuleTask
              )
            </a>
          </li>
          <li
            className=""
            role="presentation"
          >
            <a
              href="#"
              onClick={[Function]}
              onKeyDown={[Function]}
              role="menuitem"
              tabIndex="-1"
            >
              Qualify
               (
              userTask
              )
            </a>
          </li>
          <li
            className=""
            role="presentation"
          >
            <a
              href="#"
              onClick={[Function]}
              onKeyDown={[Function]}
              role="menuitem"
              tabIndex="-1"
            >
              Retract Validation
               (
              businessRuleTask
              )
            </a>
          </li>
          <li
            className=""
            role="presentation"
          >
            <a
              href="#"
              onClick={[Function]}
              onKeyDown={[Function]}
              role="menuitem"
              tabIndex="-1"
            >
              Validation1
               (
              businessRuleTask
              )
            </a>
          </li>
        </ul>
      </div>
    </div>
    <div
      className="col-sm-4"
    >
      <button
        className="btn btn-primary"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        Map these two nodes
      </button>
    </div>
  </div>
  <div
    className="form-group"
  >
    <label
      className="col-sm-4 control-label"
      htmlFor="mappingLabels"
    >
      Mappings:
    </label>
    <div
      className="col-sm-8"
    />
  </div>
  <div
    className="form-group"
  >
    <div
      className="text-right col-sm-4"
    >
      <button
        className="btn btn-default"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        Show
         Source Diagram
      </button>
    </div>
    <div
      className="text-right col-sm-4"
    >
      <button
        className="btn btn-default"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        Show
         Target Diagram
      </button>
    </div>
  </div>
</form>
`;
